persistence/remote/src/main/java/org/infinispan/persistence/remote/RemoteStore.java

```diff
@@ -299,65 +332,61 @@ public void write(MarshallableEntry entry) throws PersistenceException {
       Object key = getKey(entry);
       Object value = getValue(entry);
 
-      remoteCache.put(key, value, toSeconds(lifespan, entry.getKey(), LIFESPAN), TimeUnit.SECONDS,
-            toSeconds(maxIdle, entry.getKey(), MAXIDLE), TimeUnit.SECONDS);
+      return remoteCache.putAsync(key, value, toSeconds(lifespan, entry.getKey(), LIFESPAN), TimeUnit.SECONDS,
+            toSeconds(maxIdle, entry.getKey(), MAXIDLE), TimeUnit.SECONDS)
+            .thenApply(CompletableFutures.toNullFunction());
    }
 
    private Object getKey(MarshallableEntry entry) {
-      Object key = entry.getKey();
-      if (key instanceof WrappedByteArray)
-         return ((WrappedByteArray) key).getBytes();
-      return key;
+      return unwrap(entry.getKey());
    }
 
    private Object getValue(MarshallableEntry entry) {
       if (configuration.rawValues()) {
-         Object value = entry.getValue();
-         return value instanceof WrappedByteArray ? ((WrappedByteArray) value).getBytes() : value;
+         return unwrap(entry.getValue());
       }
       return entry.getMarshalledValue();
    }
 
    @Override
-   public CompletionStage<Void> bulkUpdate(Publisher<MarshallableEntry<? extends K, ? extends V>> publisher) {
-      return Flowable.fromPublisher(publisher)
-            .buffer(configuration.maxBatchSize())
-            // TODO: this is blocking - when the full non blocking SPI is introduced this needs to chain returns
-            // of the putAll to ensure that only a single batch is sent at a time
-            .doOnNext(entries -> remoteCache.putAll(entries.stream().collect(Collectors.toMap(this::getKey, this::getValue))))
-            .doOnError(PersistenceException::new)
-            .ignoreElements()
+   public CompletionStage<Void> batch(int publisherCount, Publisher<SegmentedPublisher<Object>> removePublisher,
+         Publisher<SegmentedPublisher<MarshallableEntry<K, V>>> writePublisher) {
+      Completable removeCompletable = Flowable.fromPublisher(removePublisher)
+            .flatMap(sp -> Flowable.fromPublisher(sp), publisherCount)
```

Comment:
`sp -> Flowable.fromPublisher` can be replaced with method references throughout.

Reply:
👍